metrics/e2e_test.go

```diff
@@ -0,0 +1,629 @@
+		init     func(t *testing.T) error
+		validate func(t *testing.T)
+	}{{
+		name: "Prometheus",
+		init: func(t *testing.T) error {
+			if err := UpdateExporter(context.Background(), configForBackend(prometheus), logtesting.TestLogger(t)); err != nil {
+				return err
+			}
+			// Wait for the webserver to actually start serving metrics
+			return wait.PollImmediate(10*time.Millisecond, 10*time.Second, func() (bool, error) {
+				resp, err := http.Get(fmt.Sprintf("http://localhost:%d/metrics", prometheusPort))
+				return err == nil && resp.StatusCode == http.StatusOK, nil
+			})
+		},
+		validate: func(t *testing.T) {
+			metricstest.EnsureRecorded()
+			resp, err := http.Get(fmt.Sprintf("http://localhost:%d/metrics", prometheusPort))
+			if err != nil {
+				t.Fatalf("failed to fetch prometheus metrics: %+v", err)
+			}
+			defer resp.Body.Close()
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				t.Fatalf("failed to read prometheus response: %+v", err)
+			}
+			want := `# HELP testComponent_global_export_counts Count of exports via standard OpenCensus view.
+# TYPE testComponent_global_export_counts counter
+testComponent_global_export_counts 2
+# HELP testComponent_resource_global_export_count Count of exports via RegisterResourceView.
+# TYPE testComponent_resource_global_export_count counter
+testComponent_resource_global_export_count 2
+# HELP testComponent_testing_value Test value
+# TYPE testComponent_testing_value gauge
+testComponent_testing_value{project="p1",revision="r1"} 0
+testComponent_testing_value{project="p1",revision="r2"} 1
+`
+			if diff := cmp.Diff(want, string(body)); diff != "" {
+				t.Errorf("Unexpected prometheus output (-want +got):\n%s", diff)
+			}
+		},
+	}, {
+		name: "OpenCensus",
+		init: func(t *testing.T) error {
+			if err := ocFake.start(len(resources) + 1); err != nil {
+				return err
+			}
+			t.Log("Created exporter at", ocFake.address)
+			return UpdateExporter(context.Background(), configForBackend(openCensus), logtesting.TestLogger(t))
+		},
+		validate: func(t *testing.T) {
+			// We unregister the views because this is one of two ways to flush
+			// the internal aggregation buffers; the other is to have the
+			// internal reporting period duration tick, which is at least
+			// [new duration] in the future.
+			metricstest.EnsureRecorded()
+			records := []metricExtract{}
+			keys := map[string]struct{}{}
+		loop:
+			for {
+				select {
+				case record := <-ocFake.published:
+					if record == nil {
+						continue loop
+					}
+					for _, m := range record.Metrics {
+						if len(m.Timeseries) > 0 {
+							labels := map[string]string{}
+							if record.Resource != nil {
+								labels = record.Resource.Labels
+							}
+							metric := metricExtract{
+								Name:   m.MetricDescriptor.Name,
+								Labels: labels,
+								Value:  m.Timeseries[0].Points[0].GetInt64Value(),
+							}
+							records = append(records, metric)
+							keys[metric.Key()] = struct{}{}
```

Comment:
Could you add a comment that why using a set here fixes the problem?

Reply:
Done. I was convinced for a long time that the RPCs weren't actually going to the right place, but I finally figured out that we simply weren't reading enough off the channel to find them.